source/common/router/router.cc

```diff
@@ -211,18 +211,24 @@ Http::FilterHeadersStatus Filter::decodeHeaders(Http::HeaderMap& headers, bool e
   }
 
   // Determine if there is a direct response for the request.
-  if (route_->directResponseEntry()) {
-    auto response_code = route_->directResponseEntry()->responseCode();
+  const auto* direct_response = route_->directResponseEntry();
+  if (direct_response != nullptr) {
+    auto response_code = direct_response->responseCode();
     if (response_code >= Http::Code::MultipleChoices && response_code < Http::Code::BadRequest) {
       config_.stats_.rq_redirect_.inc();
-      Http::Utility::sendRedirect(*callbacks_, route_->directResponseEntry()->newPath(headers),
-                                  response_code);
+      Http::Utility::sendRedirect(*callbacks_, direct_response->newPath(headers), response_code);
```

Comment:
Is there a reason why 3xx responses don't get extra headers and response body?

Reply:
That's for backward compatibility. The implementations of the new direct-response feature and the existing redirect feature were merged in PR2335, at the request of the reviewers. Thus, if we were to send the extra headers with 3xx responses, it could change the behavior of users' existing Envoy configurations.

Reply:
Suggestion here (without full review): I still think we could probably pull the trigger and just do it in all cases and no one would notice, but if we feel that we shouldn't do that, I would put the fact that we are deprecating the old behavior in DEPRECATED.md and put a comment here. We can remove the guard in the next release.

Reply:
Hmm, sending the headers and response body in all cases would definitely yield a more consistent feature. I'll update this PR to make 3xx work the same way as all other response statuses.